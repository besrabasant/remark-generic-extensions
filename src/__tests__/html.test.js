import { transformToHtml } from "./helpers/transform";

describe("html inline extensions", () => {
  test("should add an extension", () =>
    expect(transformToHtml("!icon")).toMatchSnapshot());

  test("should add an extension with an id", () =>
    expect(transformToHtml("!icon{ #my-id }")).toMatchSnapshot());

  test("should add an extension with a class", () =>
    expect(transformToHtml("!icon{ .my-class }")).toMatchSnapshot());

  test("should add an extension with an attribute", () =>
    expect(transformToHtml("!icon{ attr=my-attr }")).toMatchSnapshot());

  test("should add an extension with a quoted attribute", () =>
    expect(transformToHtml(`!icon{ attr="my attribute" }`)).toMatchSnapshot());

  test("should add an extension with a lone attribute", () =>
    expect(transformToHtml(`!icon{ attr }`)).toMatchSnapshot());

  test("should add an extension with a lone attribute with a default value", () =>
    expect(
      transformToHtml(`!icon{ attr }`, {
        elements: {
          icon: {
            propsDefaultValues: {
              attr: "default"
            }
          }
        }
      })
    ).toMatchSnapshot());

  test("should add an extension with a lone attribute with a default boolean value", () =>
    expect(
      transformToHtml(`!icon{ highlight }`, {
        elements: {
          icon: {
            propsDefaultValues: {
              highlight: true
            }
          }
        }
      })
    ).toMatchSnapshot());

  test("should add an extension with all types of parameters", () =>
    expect(
      transformToHtml(
        `!icon{ #my-id .my-class attr attr2=my-attr attr3="my attribute" }`,
        {
          elements: {
            icon: {
              propsDefaultValues: {
                attr: "default"
              }
            }
          }
        }
      )
    ).toMatchSnapshot());

  test("should add an extension with a boolean property", () =>
    expect(
      transformToHtml("!checkbox", {
        elements: {
          checkbox: {
            html: {
              tagName: "checkbox",
              properties: {
                checked: true
              }
            }
          }
        }
      })
    ).toMatchSnapshot());

  test("should add an extension with a mapped content", () =>
    expect(
      transformToHtml("!icon[my-tooltip]", {
        elements: {
          icon: {
            html: {
              properties: {
                tooltip: "::content::"
              }
            }
          }
        }
      })
    ).toMatchSnapshot());

  test("should add an extension with a mapped argument", () =>
    expect(
      transformToHtml("!icon(my-icon)", {
        elements: {
          icon: {
            html: {
              properties: {
                icon: "::argument::"
              }
            }
          }
        }
      })
    ).toMatchSnapshot());

  test("should work with a different placeholder affix", () =>
    expect(
      transformToHtml("!icon(my-icon)", {
        placeholderAffix: "||",
        elements: {
          icon: {
            html: {
              properties: {
                icon: "||argument||"
              }
            }
          }
        }
      })
    ).toMatchSnapshot());

  test("should add an extension with a replaced tag", () =>
    expect(
      transformToHtml("!icon", {
        elements: {
          icon: {
            html: {
              tagName: "my-icon"
            }
          }
        }
      })
    ).toMatchSnapshot());

  test("should work with other markdown", () =>
    expect(
      transformToHtml(
        `#heading

!icon`
      )
    ).toMatchSnapshot());

  test("should work with an element with multiple children", () =>
    expect(
      transformToHtml("!icon", {
        elements: {
          icon: {
            html: {
              children: [
                {
                  tagName: "p",
                  children: [
                    {
                      type: "text",
                      value: "test"
                    }
                  ]
                }
              ]
            }
          }
        }
      })
    ).toMatchSnapshot());

  test("should work with an element with a mapped property on a child", () =>
    expect(
      transformToHtml("!icon{ #test }", {
        elements: {
          icon: {
            html: {
              children: [
                {
                  tagName: "p",
                  properties: {
                    id: "::prop::id::"
                  }
                }
              ]
            }
          }
        }
      })
    ).toMatchSnapshot());

  test("should work with an element with a mapped content on a child value", () =>
    expect(
      transformToHtml("!icon[my-content]", {
        elements: {
          icon: {
            html: {
              children: [
                {
                  tagName: "p",
                  children: [
                    {
                      type: "text",
                      value: "::content::"
                    }
                  ]
                }
              ]
            }
          }
        }
      })
    ).toMatchSnapshot());

  test("should work with an element with multiple children on the same level", () =>
    expect(
      transformToHtml("!icon", {
        elements: {
          icon: {
            html: {
              children: [
                {
                  tagName: "p",
                  children: [
                    {
                      type: "text",
                      value: "test"
                    }
                  ]
                },
                {
                  tagName: "span",
                  children: [
                    {
                      type: "text",
                      value: "bla"
                    }
                  ]
                }
              ]
            }
          }
        }
      })
    ).toMatchSnapshot());

  test("should work with the example from README", () =>
    expect(
      transformToHtml(
        `# Alpha

!alert[My message!](my subtext is rad){ #my-alert .custom-alert }

## Bravo

## Delta`,
        {
          elements: {
            alert: {
              html: {
                tagName: "span",
                children: [
                  {
                    type: "element",
                    tagName: "i",
                    properties: {
                      className: "fa fa-exclamation",
                      ariaHidden: true
                    }
                  },
                  {
                    type: "element",
                    tagName: "span",
                    children: [
                      {
                        type: "text",
                        value: "::content::"
                      }
                    ]
                  },
                  {
                    type: "element",
                    tagName: "span",
                    properties: {
                      className: "subtext"
                    },
                    children: [
                      {
                        type: "text",
                        value: "::argument::"
                      }
                    ]
                  }
                ]
              }
            }
          }
        }
      )
    ).toMatchSnapshot());
});
