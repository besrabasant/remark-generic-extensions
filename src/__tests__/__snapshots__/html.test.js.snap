// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`html inline extensions should add an extension 1`] = `
"<p><icon></icon></p>
"
`;

exports[`html inline extensions should add an extension with a boolean property 1`] = `
"<p><checkbox checked></checkbox></p>
"
`;

exports[`html inline extensions should add an extension with a class 1`] = `
"<p><icon class=\\"my-class\\"></icon></p>
"
`;

exports[`html inline extensions should add an extension with a lone attribute 1`] = `
"<p><icon attr=\\"\\"></icon></p>
"
`;

exports[`html inline extensions should add an extension with a lone attribute with a default boolean value 1`] = `
"<p><icon highlight=\\"true\\"></icon></p>
"
`;

exports[`html inline extensions should add an extension with a lone attribute with a default value 1`] = `
"<p><icon attr=\\"default\\"></icon></p>
"
`;

exports[`html inline extensions should add an extension with a mapped argument 1`] = `
"<p><icon icon=\\"my-icon\\"></icon></p>
"
`;

exports[`html inline extensions should add an extension with a mapped content 1`] = `
"<p><icon tooltip=\\"my-tooltip\\"></icon></p>
"
`;

exports[`html inline extensions should add an extension with a quoted attribute 1`] = `
"<p><icon attr=\\"my attribute\\"></icon></p>
"
`;

exports[`html inline extensions should add an extension with a replaced tag 1`] = `
"<p><my-icon></my-icon></p>
"
`;

exports[`html inline extensions should add an extension with all types of parameters 1`] = `
"<p><icon attr3=\\"my attribute\\" attr2=\\"my-attr\\" class=\\"my-class\\" id=\\"my-id\\" attr=\\"default\\"></icon></p>
"
`;

exports[`html inline extensions should add an extension with an attribute 1`] = `
"<p><icon attr=\\"my-attr\\"></icon></p>
"
`;

exports[`html inline extensions should add an extension with an id 1`] = `
"<p><icon id=\\"my-id\\"></icon></p>
"
`;

exports[`html inline extensions should work with a different placeholder affix 1`] = `
"<p><icon icon=\\"my-icon\\"></icon></p>
"
`;

exports[`html inline extensions should work with an element with a mapped content on a child value 1`] = `
"<p><icon><p>my-content</p></icon></p>
"
`;

exports[`html inline extensions should work with an element with a mapped property on a child 1`] = `
"<p><icon><p id=\\"test\\"></p></icon></p>
"
`;

exports[`html inline extensions should work with an element with multiple children 1`] = `
"<p><icon><p>test</p></icon></p>
"
`;

exports[`html inline extensions should work with an element with multiple children on the same level 1`] = `
"<p><icon><p>test</p><span>bla</span></icon></p>
"
`;

exports[`html inline extensions should work with other markdown 1`] = `
"<p>#heading</p>
<p><icon></icon></p>
"
`;

exports[`html inline extensions should work with the example from README 1`] = `
"<h1>Alpha</h1>
<p><span class=\\"custom-alert\\" id=\\"my-alert\\"><i class=\\"fa fa-exclamation\\" aria-hidden=\\"true\\"></i><span>My message!</span><span class=\\"subtext\\">my subtext is rad</span></span></p>
<h2>Bravo</h2>
<h2>Delta</h2>
"
`;
